File: cocosjs/src/scene/LoginScene.js

var LoginLayer = cc.Layer.extend({
    _ccsNode:null,
    ctor:function () {
        this._super();
        this.createCCSNode("uiccs/LoginLayer.csb");
        var ccsNode=this._ccsNode;

        this._nameTextField=ccsNode.getChildByName("_nameTextField");
        this._psTextField=ccsNode.getChildByName("_psTextField");
        this._ipTextField=ccsNode.getChildByName("_ipTextField");

        function restoreData(key,defaultValue,textField){
            var userName =sys.localStorage.getItem(key);
            if(!!userName && userName.length>0)
            {
                textField.setString(userName);
            }
            else
            {
                sys.localStorage.setItem(key,defaultValue);
                textField.setString(defaultValue);
            }
        }
        restoreData("userName","",this._nameTextField);
        restoreData("password","",this._psTextField);
        restoreData("ip","0.0.0.0",this._ipTextField);

        var _loginButton=ccsNode.getChildByName("_loginButton");
        _loginButton.addTouchEventListener(this.touchEvent,this);
        _loginButton.setPressedActionEnabled(true);

        cc.log("LoginLayer=======>>"+this._nameTextField.getString());
        ///////////////////////////////////////
    },

    createCCSNode:function(ccbFileName){
        var x = cc.winSize.width/2;
        var y = cc.winSize.height/2;
        var ccsNode = ccs.CSLoader.createNode(ccbFileName);
        ccsNode.setPosition(x,y);
        this.addChild(ccsNode);
        this._ccsNode=ccsNode;
    },

    touchEvent: function (sender, type) {
        switch (type) {
            case ccui.Widget.TOUCH_ENDED:
                this.loginGame();
        }
    },

    loginGame:function(){
        var username=this._nameTextField.getString();
        var password=this._psTextField.getString();

        sys.localStorage.setItem("userName",username);
        sys.localStorage.setItem("password",password);

        var ip=this._ipTextField.getString();
        if (ip!=="0.0.0.0") {
            sys.localStorage.setItem("ip",ip);
            cb.CommonLib.setServerURL(ip);
        }
        
        if (this.callback) {
            this.callback(username,password);
        }
    },

    addEventListener:function(callback){
        this.callback=callback;

        if(cc.sys.os == cc.sys.OS_OSX){
            this.loginGame();
        }
    }
});

var LoginScene = cc.Scene.extend({
    ctor:function () {
        this._super();
        var layer = new LoginLayer();
        this.addChild(layer);
        this.setTag(16881);
        this.layer=layer;
    },

    addEventListener:function(callback){
        this.layer.addEventListener(callback);
    }
});

File: cocosjs/frameworks/runtime-src/Classes/GameCore/LoginScene.cpp
#include "LoginScene.h"
#include "CommonLib.h"

LoginScene::LoginScene()
{
}

LoginScene::~LoginScene()
{
    CommonLib::removeRes("uiimg/app_login.plist");
}

LoginScene* LoginScene::create()
{
    LoginScene *pRet = new LoginScene();
    if (pRet!=nullptr)
    {
        pRet->init();
        pRet->autorelease();
        return pRet;
    }
    else
    {
        delete pRet;
        pRet = NULL;
        return NULL;
    }
}

bool LoginScene::init()
{
    Node* ccsNode=CSLoader::createNode("uiccs/LoginLayer.csb");
    Size winSize=Director::getInstance()->getWinSize();
    ccsNode->setPosition(winSize.width/2,winSize.height/2);
    addChild(ccsNode);
    
    _nameTextField=(ui::TextField*)ccsNode->getChildByName("_nameTextField");
    _psTextField=(ui::TextField*)ccsNode->getChildByName("_psTextField");
    _ipTextField=(ui::TextField*)ccsNode->getChildByName("_ipTextField");
    
    std::function<void(const char*,const char*,ui::TextField*)> restoreData;
    restoreData=[](const char* key,const char* defaultValue,ui::TextField* textField){
        std::string userName=UserDefault::getInstance()->getStringForKey(key);
        if (userName.length()==0)
        {
            UserDefault::getInstance()->setStringForKey(key,defaultValue);
            textField->setString(defaultValue);
        }
        else
        {
            textField->setString(userName);
        }
    };

    restoreData("userName","",_nameTextField);
    restoreData("password","",_psTextField);
//    restoreData("ip","0.0.0.0",_ipTextField);

    ui::Button* _loginButton=(ui::Button*)ccsNode->getChildByName("_loginButton");
    _loginButton->setPressedActionEnabled(true);
    _loginButton->addTouchEventListener([=](Ref* sender,ui::Widget::TouchEventType eventType){
        if(eventType==ui::Widget::TouchEventType::ENDED)
        {
            this->loginGame();
        }
    });
    
	return true;
}

void LoginScene::loginGame()
{
    std::string username=_nameTextField->getString();
    std::string password=_psTextField->getString();
    UserDefault::getInstance()->setStringForKey("userName",username);
    UserDefault::getInstance()->setStringForKey("password",password);
//    std::string ip=_ipTextField->getString();
//    if (ip!="0.0.0.0")
//    {
//        UserDefault::getInstance()->setStringForKey("ip",ip);
//        CommonLib::setServerURL(ip.c_str());
//    }
    if (_callback)
    {
        _callback(username,password);
    }
}

void LoginScene::addEventListener(std::function<void(std::string,std::string)> callback)
{
    _callback=callback;
}


